### PR TITLE
Base Analysis Module Channelizer (WIP)

### PR DESCRIPTION
**Pull request by cczhu**
Work in progress, but wanted to show you what I have so far.

The channelizer currently works if fed a Baseband stream reader.  It itself acts like a stream on top of the Baseband stream, and defines its own channelized complete sample shape as:
```
(n_fourier_chan,) + fh.sample_shape
```
where the number of Fourier channels is user-defined.  It also defines a start time, stop time, sample shape, ndim and shape all in accordance with the channelized sample definition.

In the backend, it utilises a block system in the same way Baseband stream readers use frames: whenever a new block is reached a (currently user-defined) number of channelized samples is calculated.  These channelized samples are held in `self._block` to be used as needed by the user, until the user seeks out of the block's bounds.  Timing tests suggest a negligible change in performance if we change the size of the block, so for the channelizer the natural block size is 1 complete channelized sample.  This may change for, e.g., dedispersion.

### CODE REVIEW

File: scintillometry/channelize/base.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+from ..fourier import get_fft_maker
+
+
+__all__ = ['ChannelizeModule']
+
+
+class ChannelizeCore(object):
+
+    # NOTE: maybe force users to pass their own FFT.
+    # NOTE: always assumes time axis is 1.  Probably not a problem.
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: It's actually the second axis, or `axis=1`.  I've been doing all of my FFTs on time-domain data that looks like `(nsample,) + sample_shape`, which I first transform to `(nbigsample, ntransform) + sample_shape` to get `(nsample, nchan) + sample_shape`.

File: scintillometry/base/base.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import operator
+import numpy as np
+import astropy.units as u
+
+
+class ModuleBase(object):
+
+    def __init__(self, ih):
+        self.ih = ih
+        if self.ih.tell():
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/base/base.py]: When I was doing testing I would first load `fh`, seek around and read some samples (for some controls for my testing), then pass `fh` into the channelizer.  I imagine in a production setting someone may first try to do `am = channelizer(fh)`, and play around with that, before doing, say, `wf = squarer(binner(am))`, so this is a guard against getting unforeseen results.

[the remark this answers is not in the document]
Comment (mhvk) [reply scintillometry/base/base.py]: I don't understand why that would be a problem: the moment you read anything from `self`, it will seek in `ih`

[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/base/base.py]: OH, I just realized why I had an issue - I was using `fh` after using `am`, and of course there's no guard for that anyway.  I'll remove this.

File: scintillometry/channelize/base.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    def channelize(self, data):
+        """Split data into frequency channels."""
+        return self.fft(data.reshape(self._input_shape))
+
+
+class ChannelizeModule(ModuleBase, ChannelizeCore):
+
+    def __init__(self, ih, nchan, samples_per_block=1, FFT=None):
+        ModuleBase.__init__(self, ih)
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: The original thought was that `ChannelizeCore` could be inherited by both `ChannelizeModule` and `DedisperseModule`, so that the dedispersion didn't have to generate multiple redundant file pointers.  I suppose like with FFTs we could just initialize a channelizer with `ChannelizeModule`, which would help if we ever also implement a polyphase filter instead (though perhaps that would have a larger minimum `samples_per_block`...).

File: scintillometry/channelize/base.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+class ChannelizeCore(object):
+
+    # NOTE: maybe force users to pass their own FFT.
+    # NOTE: always assumes time axis is 1.  Probably not a problem.
+    def __init__(self, sample_shape, dtype, nchan, samples_per_block=1,
+                 FFT=None):
+        self.nchan = operator.index(nchan)
+        self.samples_per_block = max(1, operator.index(samples_per_block))
+        FFT = get_fft_maker('numpy') if FFT is None else FFT
+        self._input_shape = (self.samples_per_block, self.nchan) + sample_shape
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: I'm hoping to reserve `sample_shape` for the output shape of any class, meaning here it would be the FFT output shape, rather than the input.

It is true that we don't really need `samples_per_block` anymore since the input shape now has it.

File: scintillometry/channelize/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply scintillometry/channelize/base.py]: Maybe an abstraction to be implemented later? In any case, it still would seem better to use an instance of it rather than inherit from it.

File: scintillometry/channelize/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply scintillometry/channelize/base.py]: Yes, that makes sense: we need `sample_shape` for the output.

File: scintillometry/channelize/base.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class ChannelizeCore(object):
+
+    # NOTE: maybe force users to pass their own FFT.
+    # NOTE: always assumes time axis is 1.  Probably not a problem.
+    def __init__(self, sample_shape, dtype, nchan, samples_per_block=1,
+                 FFT=None):
+        self.nchan = operator.index(nchan)
+        self.samples_per_block = max(1, operator.index(samples_per_block))
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: I'm not sure the user should ever actually have to set `samples_per_block`.  Certainly for the dedispersion there's a minimum value that can be auto-calculated.

File: scintillometry/channelize/base.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    def __init__(self, ih, nchan, samples_per_block=1, FFT=None):
+        ModuleBase.__init__(self, ih)
+        ChannelizeCore.__init__(self, self.ih.sample_shape, self.ih.dtype,
+                                nchan, samples_per_block=samples_per_block,
+                                FFT=FFT)
+        # NOTE: nsample is the number of output samples for the largest integer
+        # number of blocks available from the input file.
+        self._nsample = self.samples_per_block * (
+            self.ih._nsample // self.nchan // self.samples_per_block)
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: I can go either one or the other, but right now Baseband stream readers don't have a `len` property, so going to use `self.ih.shape[0]`.

File: scintillometry/channelize/base.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    @lazyproperty
+    def shape(self):
+        """Shape of the (squeezed/subset) stream data."""
+        return (self._nsample,) + self.sample_shape
+
+    @property
+    def sample_shape(self):
+        return self.fft.freq_shape[1:]
+
+    @property
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: Shape can go up there too.  Moved.

File: scintillometry/channelize/base.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    @lazyproperty
+    def stop_time(self):
+        """Time at the end of the file, just after the last sample.
+
+        See also `start_time` for the start time of the file, and `time` for
+        the time of the sample pointer's current offset.
+        """
+        # NOTE: can eventually replace this with seek_temporary block.
+        current_pos = self.ih.tell()
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: True!  Changed.

File: scintillometry/channelize/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: Made the change, and I agree it now looks much nicer!

**Comment by cczhu** [comment]
Made changes to address PR comments.  Before I write a test suite, though, I want to try making the intensity module, and maybe even the dedispersion module, to see how well our framework generalizes.

We should also think of better names.  "Module" has multiple meanings in a Python context, as does "block" for radio data analysis.  Something like "Operator" and "buffer" might make more sense?

File: scintillometry/base/base.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+
+class ModuleBase(object):
+
+    def __init__(self, ih):
+        self.ih = ih
+        self.offset = 0
+        self._block_index = None
+
+    @lazyproperty
```
Comment (mhvk) [inline scintillometry/base/base.py]: Why a `lazyproperty` for something so trivial? Actually, I think this should remain up - I don't like relying on private attributes here.

File: scintillometry/channelize/base.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+from astropy.utils import lazyproperty
+
+from ..base.base import ModuleBase
+from ..fourier import get_fft_maker
+
+
+__all__ = ['ChannelizeModule']
+
+
+class Channelizer(object):
```
Comment (mhvk) [inline scintillometry/channelize/base.py]: I still am not sure there is a point in having this separate class.... Maybe keep it simple for now?

File: scintillometry/channelize/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply scintillometry/channelize/base.py]: This still seems needlessly complicated. Why not just
```
return self.start_time + self.size / self.sample_rate
```

[the remark this answers is not in the document]
Comment (mhvk) [reply scintillometry/channelize/base.py]: In fact, even `start_time` and `stop_time` could arguably go up to the base module class... But fine to do that later too.

**Comment by mhvk** [comment]
On the names: how about just using `frame` instead of `block`, so we stick with the analogy with the reader?
For the `Module`, I agree that name is problematic. Googling `python pipeline`, I encountered the `scikit-learn` pipeline, which seems to have `Transformer` modules [1]. Alternatively, there is an astrophysics pipeline package which uses `PipelineElement` (bit verbose, really) [2]

[1] https://stackoverflow.com/questions/33091376/python-what-is-exactly-sklearn-pipeline-pipeline#33094099
[2] https://pythonhosted.org/Astropysics/coremods/pipeline.html

**Comment by mhvk** [comment]
On tests: I think it would be best to have at least a few very simple one, that just check that functionality is correct (but don't worry much about attributes/properties/etc.)

File: scintillometry/channelize/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: Point taken.  I've also moved both to `base/base`, since they do make good defaults.

File: scintillometry/base/base.py (hunk 9 above)
Comment (cczhu) [reply scintillometry/base/base.py]: I think this is the case for all the lazy properties in base (none require extensive calculation), so changed them all to properties.

File: scintillometry/channelize/base.py (hunk 10 above)
Comment (cczhu) [reply scintillometry/channelize/base.py]: The original thought was `Channelizer` could also be used in `DedisperseModule`, but largely agreed this is an unnecessary abstraction for now (especially since the actual core is the `fft`, which is already abstracted).  Changed.

**Comment by cczhu** [comment]
I don't mind using `frame` for now, even though it feels like there should be metadata associated with a data block if it's to be called frame.  For renaming `Module`, I think `Operation` is a better name than either `Transformer` or `Element` (at least because `DedispersionOperation` sounds better than `DedisperseTransformer` or something).  That said, celery[1], dask[2] and luigi[3] all use "task" to describe one element of a workflow, and I don't mind `ChannelizationTask`, `DedispersionTask`, etc.  The only caveat is that a "task" in luigi is its own class, and using the same name may make the code confusing to read.  For now, though, I'm going to use "Task" because it's short and to-the-point.

(Though, maybe "Step" or "Stage" would make even more sense?)

Tests are forthcoming - will be done in a few hours.

[1] http://docs.celeryproject.org/en/latest/getting-started/first-steps-with-celery.html#first-steps-with-celery
[2] https://dask.pydata.org/en/latest/custom-graphs.html#custom-graph-example
[3] https://github.com/spotify/luigi#background

**Comment by cczhu** [comment]
Tests are now done.

**Comment by mhvk** [comment]
p.s. All this just goes to show that seeing real code really helps to bring into focus what is and what is not needed!

**Comment by cczhu** [comment]
I'm not sure I follow regarding `samples_per_frame`.  If pyfftw is used, a fixed FFT size is required, which is why we select a fixed frame (if we don't want this to be true I'm going to have to revise the way the FFT module works, and we lose some of the speed afforded by pyfftw).  In the absence of any information about subsequent tasks, I don't see how to set a maximum number of samples.

Alternatively, we can set `samples_per_frame=1` always.  I did some profiling that suggests it's not a significant overhead to the actual FFT.

By the way, I currently have some changes to the Fourier module as well, to make using it more convenient for the channelizer.  I can separate these out into a separate PR.

File: scintillometry/channelize/tests/test_channelize.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+        ct.seek(-2, 2)
+        with pytest.raises(EOFError):
+            ct.read(10)
+
+        ct.close()
+        assert fh.closed
+
+    @pytest.mark.parametrize('spf', (5, 8))
+    def test_channelizeblocksize(self, spf):
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/tests/test_channelize.py]: Regardless of what we end up doing with `samples_per_frame` for channelize, I'm deleting this in favour of a `test_base.py` in `scintillometry/base/tests`.  The only reason this currently exists is to test that frames are consistently defined, but that's silly to do for a task that cares little about this.

File: scintillometry/channelize/base.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            self.ih.shape[0] // self.nchan // self.samples_per_frame)
+        assert self._nsample > 0, ("not enough samples to fill one frame of "
+                                   "data!")
+        self.sample_rate = ih.sample_rate / self.nchan
+
+        # Initialize channelizer.
+        if FFT is None:
+            FFT = get_fft_maker('numpy')
+        self._fft = FFT(
+            (self.samples_per_frame, self.nchan) + self.ih.sample_shape,
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/base.py]: As implied by my comments below, PyFFTW works with arrays of fixed size, so the FFT module does as well.  There's currently no capability to pass an array where the axes not being transformed can have arbitrary length.

File: scintillometry/channelize/tests/test_channelize.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    def test_channelizetask(self):
+        """Test channelization task."""
+
+        fh = vdif.open(SAMPLE_VDIF)
+        ct = ChannelizeTask(fh, self.nchan)
+
+        # Check sample pointer.
+        assert ct.sample_rate == self.ref_sample_rate / self.nchan
+        assert ct.shape == self.ref_data.shape
+        assert ct.size == np.prod(self.ref_data.shape)
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/channelize/tests/test_channelize.py]: Changed.

File: scintillometry/base/base.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+    def __init__(self, ih):
+        self.ih = ih
+        self.offset = 0
+        self._frame_index = None
+
+    @property
+    def shape(self):
+        """Shape of the (squeezed/subset) stream data."""
+        return (self._nsample,) + self.sample_shape
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/base/base.py]: I'm not too happy about it either way, but for now will make `shape` a positional argument.

Doesn't Baseband use `self._nsample` when calculating its shape?

File: scintillometry/base/base.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import operator
+import numpy as np
+import astropy.units as u
+
+
+class TaskBase(object):
```
[the remark this answers is not in the document]
Comment (cczhu) [reply scintillometry/base/base.py]: Agreed that values should be passed if `TaskBase` methods need them (otherwise it just gets confusing).  I've made `shape`, `sample_rate` and `dtype` positional arguments to `__init__` - this way we're guaranteed for eg. the `sample_shape` property to work.

**Comment by cczhu** [comment]
(Addressed all minor comments, but test suite is still incomplete, and I'm still not certain how to fix the size of the channelizer's frame.)

**Comment by mhvk** [comment]
@cczhu - The problem is simply that I keep forgetting about `pyfftw`! But if you didn't find much of an overhead, I think we should just set `samples_per_frame=1` for now; I'd like to keep things simple...

**Comment by cczhu** [comment]
@mhvk I've moved the `samples_per_frame` testing from the channelization module to the base one, since there's currently no reason to alter `samples_per_frame` for the channelizer.  `samples_per_frame` remains an optional parameter to `ChannelizeTask`, as I don't think the question of exactly how to set the batch size (maybe `samples_per_batch` is a better name?) will be resolved until we have a working pipeline.

**Comment by cczhu** [comment]
Of course, I made the stupid mistake of being vague about what `nchan` means for channelizing real signals.  Right now it's really the bin size when reshaping in the time domain, so the output of the channelization actually has `nchan // 2 + 1` channels.  It probably makes more sense for `nchan` to mean the number of frequency channels in the output data; for real signals, `nchan` should internally get reset to `nchan + 1`.

**Comment by cczhu** [comment]
By `nsample`, do you mean the time-domain counterpart to `nchan`, so that the user inputs `nsample` and `nchan = nsample // 2 + 1` for real FFTs?  I think that would make sense.

**Comment by mhvk** [comment]
Yes, that is what I meant

**Comment by cczhu** [comment]
Revised some naming conventions, in particular changing `nchan` to `n`.  Will do the same for the dedisperser.  Made `samples_per_frame` private in `TaskBase` (since the user shouldn't really have access to it; we can revisit this later).  Added documentation and docstrings.

If this update is accepted, I'll ``rebase -i`` and squash all the nonsense commit messages.

File: scintillometry/base/tests/test_base.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+        assert rt.size == np.prod(rt.shape)
+        assert rt.ndim == fh.ndim + 1
+        assert rt.tell() == 0
+        assert rt.tell(unit='time') == rt.time == rt.start_time
+        assert abs(rt.stop_time - rt.start_time -
+                   (nsample * n) / fh.sample_rate) < 1*u.ns
+
+        # Get reference data.
+        ref_data = fh.read(nsample * n).reshape((-1, n) + fh.sample_shape)
+        # Move fh back for rt to use.
```
Comment (mhvk) [inline scintillometry/base/tests/test_base.py]: Why is this needed?

File: scintillometry/base/base.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+            data = self._frame[sample_offset:sample_offset + nsample]
+            # Copy to relevant part of output.
+            out[sample:sample + nsample] = data
+            sample += nsample
+            self.offset = offset0 + sample
+            count -= nsample
+
+        return out
+
+    def close(self):
```
Comment (mhvk) [inline scintillometry/base/base.py]: Should we have `__enter__` and `__exit__` methods as well? I considered `closed` as well, but we do not have that in baseband either, so probably fine to omit.

File: scintillometry/base/tests/test_base.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+        # Check reading to external array.
+        out = np.empty((11,) + rt.sample_shape)
+        rt.seek(0)
+        rt.read(out=out)
+        assert np.all(out == ref_data[:11])
+
+        # Check closing.
+        rt.close()
+        assert fh.closed
```
Comment (mhvk) [inline scintillometry/base/tests/test_base.py]: ~Probably also should check that `rt.closed` holds.~ Never mind - don't have that in baseband either.

File: scintillometry/channelize/__init__.py [hunk 19]
```diff
@@ -0,0 +1,4 @@
+# Licensed under the GPLv3 - see LICENSE
+"""Channelization modules."""
+
+from .base import ChannelizeTask
```
Comment (mhvk) [inline scintillometry/channelize/__init__.py]: Refactoring into a single `channelize.py` task is kept separate?

File: scintillometry/channelize/base.py [hunk 20]
```diff
@@ -0,0 +1,65 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import operator
+
+from ..base.base import TaskBase
```
Comment (mhvk) [inline scintillometry/channelize/base.py]: Similarly, a bit many `base` - refactoring to be done after?

File: scintillometry/channelize/base.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+        FFT maker.  Default: `None`, in which case the channelizer uses
+        `~scintillometry.fourier.numpy.NumpyFFTMaker`.
+
+    Notes
+    -----
+    Instances initialize an FFT that acts upon axis 1 of an input with shape::
+
+        (samples_per_frame, n) + ih.sample_shape
+
+    Setting ``samples_per_frame`` to a number larger than 1 results in the FFT
```
Comment (mhvk) [inline scintillometry/channelize/base.py]: Nice to have this note!

File: scintillometry/channelize/tests/test_channelize.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+            self.ref_start_time = fh.start_time
+            self.ref_sample_rate = fh.sample_rate
+            data = fh.read()
+
+        last_sample = self.n * (data.shape[0] // self.n)
+        self.ref_data = np.fft.rfft(
+            data[:last_sample].reshape((-1, self.n) + data.shape[1:]),
+            axis=1).astype('complex64')
+
+    def test_channelizetask(self):
```
Comment (mhvk) [inline scintillometry/channelize/tests/test_channelize.py]: Maybe check with both `pyfftw` and `numpy`? Easiest might be to use `pyfftw` if available.

File: scintillometry/channelize/base.py [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+        samples_per_frame = operator.index(samples_per_frame)
+        sample_rate = ih.sample_rate / n
+
+        nsample = samples_per_frame * (ih.shape[0] // n // samples_per_frame)
+        assert nsample > 0, "not enough samples to fill one frame of data!"
+        self._raw_frame_len = n * samples_per_frame
+
+        # Initialize channelizer.
+        if FFT is None:
+            FFT = get_fft_maker('numpy')
```
Comment (mhvk) [inline scintillometry/channelize/base.py]: Probably for later, but this probably should get a "default FFT", which is a system setting with a default (for now) of `pyfftw` if available, otherwise `numpy`.

File: scintillometry/base/tests/test_base.py (hunk 16 above)
Comment (cczhu) [reply scintillometry/base/tests/test_base.py]: It's not!  Removed.

File: scintillometry/base/base.py (hunk 17 above)
Comment (cczhu) [reply scintillometry/base/base.py]: I'm not a huge fan of writing something like:
```
with channelize(dedisperse(vdif.open('myfile.vdif'))) as ch:
    ...
```
It would work, but it's really ugly and I'd rather not the user get acquainted with writing things this way.  I almost think the `__enter__` and `__exit__` should be part of the task governor that uses sciluigi or something to manage the tasks.

File: scintillometry/channelize/__init__.py (hunk 19 above)
Comment (cczhu) [reply scintillometry/channelize/__init__.py]: My plan is to have this go through, then refactor the entire code as we discussed.

File: scintillometry/channelize/base.py (hunk 20 above)
Comment (cczhu) [reply scintillometry/channelize/base.py]: As above.

File: scintillometry/channelize/base.py (hunk 23 above)
Comment (cczhu) [reply scintillometry/channelize/base.py]: I'd do that later.

File: scintillometry/channelize/tests/test_channelize.py (hunk 22 above)
Comment (mhvk) [reply scintillometry/channelize/tests/test_channelize.py]: Fair enough - though obviously the test setup could use the same FFT routine. The real question here is testing the setting of the FFT. Anyway, am happy to leave that out for now.

File: scintillometry/channelize/base.py (hunk 23 above)
Comment (mhvk) [reply scintillometry/channelize/base.py]: OK, new issue raised.

File: scintillometry/base/base.py (hunk 17 above)
Comment (mhvk) [reply scintillometry/base/base.py]: OK, raised new issue to remind us (me...)
